pallets/asset/src/lib.rs

```diff
@@ -2244,4 +2305,81 @@ impl<T: Trait> Module<T> {
                 .unwrap_or(ERC1400_TRANSFER_FAILURE),
         )
     }
+
+    /// Transfers an asset using custodial allowance
+    fn unsafe_transfer_by_custodian(
+        custodian_did: IdentityId,
+        ticker: Ticker,
+        holder_did: IdentityId,
+        receiver_did: IdentityId,
+        value: T::Balance,
+    ) -> DispatchResult {
+        let mut custodian_allowance =
+            Self::custodian_allowance((ticker, holder_did, custodian_did));
+        // using checked_sub (safe math) to avoid underflow
+        custodian_allowance = custodian_allowance
+            .checked_sub(&value)
+            .ok_or(Error::<T>::AllowanceUnderflow)?;
+        // using checked_sub (safe math) to avoid underflow
+        let new_total_allowance = Self::total_custody_allowance((ticker, holder_did))
+            .checked_sub(&value)
+            .ok_or(Error::<T>::TotalAllowanceUnderflow)?;
+        // Validate the transfer
+        ensure!(
+            Self::_is_valid_transfer(
+                &ticker,
+                <identity::Module<T>>::did_records(custodian_did).master_key,
+                Some(holder_did),
+                Some(receiver_did),
+                value
+            )? == ERC1400_TRANSFER_SUCCESS,
+            Error::<T>::InvalidTransfer
+        );
+        Self::unsafe_transfer(custodian_did, &ticker, holder_did, receiver_did, value)?;
+        // Update Storage of allowance
+        <CustodianAllowance<T>>::insert((ticker, holder_did, custodian_did), &custodian_allowance);
+        <TotalCustodyAllowance<T>>::insert((ticker, holder_did), new_total_allowance);
+        Self::deposit_event(RawEvent::CustodyTransfer(
+            custodian_did,
+            ticker,
+            holder_did,
+            receiver_did,
+            value,
+        ));
+        Ok(())
+    }
+
+    /// Internal function to process a transfer without any checks.
+    /// Used for reverting failed settlements
+    fn unsafe_system_transfer(
+        sender: IdentityId,
+        ticker: &Ticker,
+        from_did: IdentityId,
+        to_did: IdentityId,
+        value: T::Balance,
+    ) {
+        let sender_balance = Self::balance(ticker, &from_did);
+        let updated_from_balance = sender_balance.saturating_sub(value);
+        let receiver_balance = Self::balance(ticker, &to_did);
+        let updated_to_balance = receiver_balance.saturating_add(value);
+
+        Self::_update_checkpoint(ticker, from_did, sender_balance);
+        Self::_update_checkpoint(ticker, to_did, receiver_balance);
+
+        // reduce sender's balance
+        <BalanceOf<T>>::insert(ticker, &from_did, updated_from_balance);
+
+        // increase receiver's balance
+        <BalanceOf<T>>::insert(ticker, &to_did, updated_to_balance);
```

Comment:
Here, if `to_did == from_did` then it will affect the calculation. The same bug pointed out by the Pablo.

Reply:
I'll add a check in the settlement module to make sure this never happens. I don't want this function to revert in any case so I'll let this be like this.